Introduce lib-embedded-gcov

michpappas opened the pull request:
Port of embedded-gcov[1] for Unikraft.

[1] https://github.com/nasa-jpl/embedded-gcov

Commit:
Introduce lib-embedded-gcov

Port of embedded-gcov[1] for Unikraft. This patch only provides
support for serial console output.

[1] https://github.com/nasa-jpl/embedded-gcov

Signed-off-by: Michalis Pappas <michalis@unikraft.org>

Starnox commented on Makefile.uk:
```diff
@@ -0,0 +1,78 @@
+endif
+
+LIBEMBEDDEDGCOV_CFLAGS-$(CONFIG_LIBEMBEDDEDGCOV_OUTPUT_SERIAL_HEXDUMP)     += -DGCOV_OPT_OUTPUT_SERIAL_HEXDUMP
+LIBEMBEDDEDGCOV_CFLAGS-$(CONFIG_LIBEMBEDDEDGCOV_OUTPUT_SERIAL_HEXDUMP)     += -Dgcov_printf=printf
```
The gcov_printf function is not only used when the `CONFIG_LIBEMBEDDEDGCOV_OUTPUT_SERIAL_HEXDUMP` option is set, but also when the `GCOV_OPT_PRINT_STATUS` option is set. This can be seen in the original embedded-gcov code where the gcov_printf function is used when either of these options is defined. This means that even if the user selects a memory output, they may still want to see console information about the gcov run.

```
#if defined(GCOV_OPT_PRINT_STATUS) || defined(GCOV_OPT_OUTPUT_SERIAL_HEXDUMP)
    GCOV_PRINT_STR("Gcov End");
    GCOV_PRINT_STR("\n");
#endif
```

```suggestion
LIBEMBEDDEDGCOV_CFLAGS-y												   += -Dgcov_printf=printf
```

michpappas replied:
Thanks, updated 👍🏼

DeliaPavel reviewed (approved):
Reviewed-by: Delia Pavel <delia_maria.pavel@stud.acs.upb.ro>

razvand commented:
@michpappas, @Starnox, I get the error below when testing `lib-embedded-gcov`. I'm running the basic [`app-helloworld` app](https://github.com/unikraft/app-helloworld), and I added the `lib-embedded-gcov` library to it. Any idea why this is happening?

```
/media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/scripts/workdir/apps/app-helloworld/build/libembeddedgcov/origin/code/gcov_public.c: In function ‘__gcov_call_constructors’:
/media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/scripts/workdir/apps/app-helloworld/build/libembeddedgcov/origin/code/gcov_public.c:206:36: error: ‘UINT32’ undeclared (first use in this function); did you mean ‘MIN3’?
         func = (void ( *)(void))(*(UINT32 *)ctor);
                                    ^~~~~~
                                    MIN3
/media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/scripts/workdir/apps/app-helloworld/build/libembeddedgcov/origin/code/gcov_public.c:206:36: note: each undeclared identifier is reported only once for each function it appears in
/media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/scripts/workdir/apps/app-helloworld/build/libembeddedgcov/origin/code/gcov_public.c:206:44: error: expected expression before ‘)’ token
         func = (void ( *)(void))(*(UINT32 *)ctor);
```

razvand commented:
OK, that error appears when selecting `Provide call constructors` in `make menuconfig`. I had selected all items by default. Still, that issue should be solved. Even if all items are selected, there shouldn't be a build error.

Also, I'm running the commands above, and when running the processing scripts, I get:

```console
$ ../../libs/embedded-gcov/scripts/gcov_process.sh build/ serial.log
Capturing coverage data from ../objs/
Found gcov version: 7.5.0
Scanning ../objs/ for .gcno files ...
Found 68 graph files in ../objs/
Processing objs/print.gcno
geninfo: WARNING: /media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/scripts/workdir/apps/app-helloworld/build/libembeddedgcov/origin/objs/print.gcno: Overlong record
at end of file!
Processing objs/__secs_to_tm.gcno
geninfo: WARNING: /media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/scripts/workdir/apps/app-helloworld/build/libembeddedgcov/origin/objs/__secs_to_tm.gcno: Overlong
record at end of file!
Processing objs/ctx.isr.gcno
geninfo: WARNING: /media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/scripts/workdir/apps/app-helloworld/build/libembeddedgcov/origin/objs/ctx.isr.gcno: Overlong record at end of file!
Processing objs/gcov_printf.gcno
geninfo: ERROR: /media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/scripts/workdir/apps/app-helloworld/build/libembeddedgcov/origin/objs/gcov_printf.gcno: reached unexpected end of file
```

Any clue what might be causing this?

razvand commented on README.md:
```diff
@@ -1,2 +1,48 @@
+
+## Collecting Coverage Results
+
+embedded-gcov provides three types of output:
```
```suggestion
`lib-embedded-gcov` provides three types of output:

```

razvand commented on README.md:
```diff
@@ -1,2 +1,48 @@
+The console output then needs to be processed into an lcov report. `lib-embedded-gcov` provides the `gcov_process.sh` script for that, which is essentially a wrapper around the tools provided by `embedded-gcov`.
+
+Before executing the script make sure you have the required dependencies installed, that is:
+- dox2unix
```
```suggestion

- `dos2unix`
```

razvand commented on README.md:
```diff
@@ -1,2 +1,48 @@
+
+Before executing the script make sure you have the required dependencies installed, that is:
+- dox2unix
+- lcov
```
```suggestion
- `lcov`
```

razvand commented on README.md:
````diff
@@ -1,2 +1,48 @@
+
+With dependencies installed, execute the script as:
+```
+lib-embedded-gcov/scripts/gcov_process.sh <build_directory> <console_log>
````
Add a concrete run sample, it's easier to know what your command should look like.

razvand commented on README.md:
````diff
@@ -1,2 +1,48 @@
+```
+
+The script will generate an lcov report and provide a link as shown below.
+```
````
These three backticks are here by mistake, right?

razvand commented on README.md:
````diff
@@ -1,2 +1,48 @@
+
+To obtain coverage results via the console you need to dump the console output into a file. With QEMU this is possible by configuring the serial device to log all output into a file:
+```
+-chardev stdio,id=char0,logfile=serial.log,signal=off -serial chardev:char0
````
Add a concrete run sample, it's easier to know what your command should look like.

razvand replied:
Also mention NOT using `-nographic` because that clashes with the redirection of standard output.

razvand commented on README.md:
````diff
@@ -1,2 +1,48 @@
+
+lcov report in file:///home/mpp/devel/unikraft_oss/uk_embedded-gcov/app-helloworld/build/libembeddedgcov/origin/results/html/index.html
+```
+
````
Why empty line?

razvand reviewed (changes requested):
See my comments.

michpappas commented:
> @michpappas, @Starnox, I get the error below when testing `lib-embedded-gcov`. I'm running the basic [`app-helloworld` app](https://github.com/unikraft/app-helloworld), and I added the `lib-embedded-gcov` library to it. Any idea why this is happening?
> 
> ```
> /media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/scripts/workdir/apps/app-helloworld/build/libembeddedgcov/origin/code/gcov_public.c: In function ‘__gcov_call_constructors’:
> /media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/scripts/workdir/apps/app-helloworld/build/libembeddedgcov/origin/code/gcov_public.c:206:36: error: ‘UINT32’ undeclared (first use in this function); did you mean ‘MIN3’?
>          func = (void ( *)(void))(*(UINT32 *)ctor);
>                                     ^~~~~~
>                                     MIN3
> /media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/scripts/workdir/apps/app-helloworld/build/libembeddedgcov/origin/code/gcov_public.c:206:36: note: each undeclared identifier is reported only once for each function it appears in
> /media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/scripts/workdir/apps/app-helloworld/build/libembeddedgcov/origin/code/gcov_public.c:206:44: error: expected expression before ‘)’ token
>          func = (void ( *)(void))(*(UINT32 *)ctor);
> ```

@razvand thanks for that. As you pointed out it happens if you enable `CONFIG_LIBEMBEDDEDGCOV_PROVIDE_CALL_CONSTRUCTORS`. I had previously come across it and was under the impression that I fixed it. I pushed an additional patch now under `patches/`.

Please notice that even with that patch the build will fail with:
```
/usr/bin/ld: /home/mpp/devel/unikraft_oss/uk_embedded-gcov/app-helloworld/build/app-helloworld_kvm-x86_64.o: in function `__gcov_call_constructors':
/home/mpp/devel/unikraft_oss/uk_embedded-gcov/app-helloworld/build/libembeddedgcov/origin/code/gcov_public.c:203: undefined reference to `__ctor_list'
/usr/bin/ld: /home/mpp/devel/unikraft_oss/uk_embedded-gcov/app-helloworld/build/libembeddedgcov/origin/code/gcov_public.c:203: undefined reference to `__ctor_end'
/usr/bin/ld: /home/mpp/devel/unikraft_oss/uk_embedded-gcov/app-helloworld/build/libembeddedgcov/origin/code/gcov_public.c:203: undefined reference to `__ctor_end'
collect2: error: ld returned 1 exit status
make[3]: *** [/home/mpp/devel/unikraft_oss/uk_embedded-gcov/unikraft/plat/kvm/Linker.uk:35: /home/mpp/devel/unikraft_oss/uk_embedded-gcov/app-helloworld/build/app-helloworld_kvm-x86_64.dbg] Error 1
```
This is normal behavior, as with this option you need to provide `__ctors_list` and `__ctors_end`. I updated the help message in the relevant option in KConfig to make it more clear.

michpappas commented:
@razvand `README.md` has now been rewritten.
